File: frontend-mobile/flutter_application/lib/services/chatbot_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'auth_service.dart';

class ChatbotService {
  // Use http://10.0.2.2:8080 for Android emulator
  // Use http://127.0.0.1:8080 for Linux/Web
  static const String baseUrl = 'http://127.0.0.1:8080/api';
  final AuthService _authService = AuthService();

  Future<Map<String, dynamic>> tokenLogin() async {
    try {
      final token = await _authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'No authentication token found'};
      }

      final response = await http.post(
        Uri.parse('$baseUrl/auth/token-login'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({'token': token}),
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        final errorData = json.decode(response.body);
        return {
          'success': false,
          'message': errorData['detail'] ?? 'Login failed',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }

  Future<Map<String, dynamic>> sendMessage(String sessionId, String message) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/chat'),
        headers: {
          'Content-Type': 'application/json',
          'X-Session-ID': sessionId,
        },
        body: json.encode({'message': message}),
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        final errorData = json.decode(response.body);
        return {
          'success': false,
          'message': errorData['detail'] ?? 'Failed to send message',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }

  Future<Map<String, dynamic>> getConversations(String sessionId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/conversations'),
        headers: {'X-Session-ID': sessionId},
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        return {'success': false, 'message': 'Failed to fetch conversations'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }

  Future<Map<String, dynamic>> startNewChat(String sessionId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/conversations/new'),
        headers: {'X-Session-ID': sessionId},
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        return {'success': false, 'message': 'Failed to start new chat'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }

  Future<Map<String, dynamic>> loadConversation(String sessionId, String conversationId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/conversations/$conversationId/load'),
        headers: {'X-Session-ID': sessionId},
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        return {'success': false, 'message': 'Failed to load conversation'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }

  Future<Map<String, dynamic>> getConversationMessages(String sessionId, String conversationId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/conversations/$conversationId/messages'),
        headers: {'X-Session-ID': sessionId},
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        return {'success': false, 'message': 'Failed to fetch messages'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Connection error: $e'};
    }
  }
}

File: frontend-desktop/src/services/chatbotService.js
import axios from 'axios';

const chatbotApi = axios.create({
    baseURL: 'http://localhost:8080/api',
    headers: {
        'Content-Type': 'application/json',
    },
    timeout: 30000,
});

export const chatbotService = {
    /**
     * Login to the chatbot API and get a session ID.
     */
    login: async (email, password) => {
        const response = await chatbotApi.post('/auth/login', { email, password });
        return response.data;
    },

    /**
     * Login to the chatbot using a JWT token.
     */
    tokenLogin: async (token) => {
        const response = await chatbotApi.post('/auth/token-login', { token });
        return response.data;
    },

    /**
     * Send a chat message. Requires a valid session ID.
     */
    sendMessage: async (sessionId, message) => {
        const response = await chatbotApi.post(
            '/chat',
            { message },
            { headers: { 'X-Session-ID': sessionId } }
        );
        return response.data;
    },

    /**
     * Get conversation history for the current session.
     */
    getHistory: async (sessionId, limit = 50) => {
        const response = await chatbotApi.get('/history', {
            params: { limit },
            headers: { 'X-Session-ID': sessionId },
        });
        return response.data;
    },

    /**
     * Start a new chat session (resets current context).
     */
    startNewChat: async (sessionId) => {
        const response = await chatbotApi.post(
            '/conversations/new',
            {},
            { headers: { 'X-Session-ID': sessionId } }
        );
        return response.data;
    },

    /**
     * List all past conversations for the user.
     */
    getConversations: async (sessionId) => {
        const response = await chatbotApi.get('/conversations', {
            headers: { 'X-Session-ID': sessionId },
        });
        return response.data;
    },

    /**
     * Load a specific past conversation.
     */
    loadConversation: async (sessionId, conversationId) => {
        const response = await chatbotApi.post(
            `/conversations/${conversationId}/load`,
            {},
            { headers: { 'X-Session-ID': sessionId } }
        );
        return response.data;
    },

    /**
     * Get message history for a specific conversation.
     */
    getConversationMessages: async (sessionId, conversationId) => {
        const response = await chatbotApi.get(
            `/conversations/${conversationId}/messages`,
            { headers: { 'X-Session-ID': sessionId } }
        );
        return response.data;
    },

    /**
     * Logout from the chatbot and destroy the session.
     */
    logout: async (sessionId) => {
        const response = await chatbotApi.post(
            '/auth/logout',
            {},
            { headers: { 'X-Session-ID': sessionId } }
        );
        return response.data;
    },

    /**
     * Check if the chatbot API is healthy.
     */
    healthCheck: async () => {
        try {
            const response = await chatbotApi.get('/health');
            return response.data;
        } catch {
            return { status: 'unreachable', database_connected: false };
        }
    },
};

export default chatbotService;
